docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+
+`cc-shim` has an implicit knowledge about which VM agent will handle those streams and signals and thus act as
+an encapsulation layer between `containerd-shim` and the `agent`:
+
+- It fragments and encapsulates the standard input stream from containerd-shim into `agent` stream packets:
+```
+  ┌───────────────────────────┬───────────────┬────────────────────┐
+  │ IO stream sequence number │ Packet length │ IO stream fragment │
+  │         (8 bytes)         │    (4 bytes)  │                    │
+  └───────────────────────────┴───────────────┴────────────────────┘
+```
+- It de-encapsulates and assembles standard output and error `agent` stream packets into an output stream
+that it forwards to `containerd-shim`
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `agent` `KILLCONTAINER` commands
+that it sends to the VM via `cc-proxy` UNIX named socket.
+
+The IO stream sequence numbers are passed from `cc-runtime` to `cc-shim` when the former spawns the latter.
+They are generated by the `agent` and `cc-runtime` fetches them by sending the `AllocateIO` command to
+`cc-proxy`.
+
+As an example, let's say that running the `pwd` command from a container standard input will generate
+`/tmp` from the container standard output. The `agent` assigned this specific process 8888 and 8889 respectively
+as the stdin, stdout and stderr sequence numbers.
+With `cc-shim` and Clear Containers, this example would look like:
+
+![cc-shim](shim.png)
+
+#### Networking
+
+Containers will typically live in their own, possibly shared, networking namespace.
+At some point in a container lifecycle, container engines will set up that namespace
+to add the container to a network which is isolated from the host network, but which is shared between containers
+
+In order to do so, container engines will usually add one end of a `virtual ethernet
+(veth)` pair into the container networking namespace. The other end of the `veth` pair
+is added to the container network.
+
+This is a very namespace-centric approach as QEMU can not handle `veth` interfaces.
+Instead it typically creates `TAP` interfaces for adding connectivity to a virtual
+machine.
+
+To overcome that incompatibility between typical container engines expectations
+and virtual machines, `cc-runtime` networking transparently bridges `veth`
+interfaces with `TAP` ones:
+
+![Clear Containers networking](network.png)
+
+The [virtcontainers library](https://github.com/containers/virtcontainers#cnm) has some more
+details on how `cc-runtime` implements [CNM](https://github.com/docker/libnetwork/blob/master/docs/design.md).
+
+## Appendices
+
+### DAX
+
+Clear Containers utilises the Linux kernel DAX (Direct Access filesystem)
+feature to efficiently map some host side files into the guest VM space.
+In particular, Clear Containers uses the `QEMU` nvdimm feature to provide a
+memory mapped virtual device that can be used to DAX map the mini-OS root
```

Comment:
This seems to be the first mention of the mini-OS. I think it might be helpful to introduce it when we first start taking about the VM, pointing out that there is a mini-OS and the 9p-mapped docker image (`ubuntu`, `busybox`, etc).